csrc/scheduler/greedy.cpp

```diff
@@ -452,6 +590,78 @@ class ConstrainedOpScheduler : public OptOutDispatch {
     scheduleConstrainedTv(out_tv, {scan_dim});
   }
 
+  void handle(ScatterOp* scatter) override {
+    auto scatter_dim = scatter->dim();
+    auto in_tv = ir_utils::getTvInput(scatter);
+    auto index_tv = scatter->index()->as<TensorView>();
+    auto src_tv = scatter->src()->as<TensorView>();
+    auto out_tv = ir_utils::getTvOutput(scatter);
+
+    scheduleConstrainedTv(out_tv, {scatter_dim});
+    scheduleConstrainedTv(index_tv, {scatter_dim});
+    scheduleConstrainedTv(in_tv, {scatter_dim});
+    scheduleConstrainedTv(src_tv, {scatter_dim});
+
+    // If in_tv is produced by another scatter, it should be already
+    // scheduled, including its memory type. It should not be
+    // modified.
+    const bool in_tv_is_produced_by_scatter =
+        dynamic_cast<ScatterOp*>(in_tv->definition()) != nullptr;
+
+    if (in_tv_is_produced_by_scatter) {
+      auto in_tv_current_memory_type = in_tv->getMemoryType();
+      NVF_ERROR_EQ(
+          in_tv->getMemoryType(),
+          in_tv_current_memory_type,
+          "Inconsistent setting of memory type of ",
+          in_tv->toString(),
+          ". Previous: ",
+          in_tv_current_memory_type,
+          ". New: ",
+          in_tv->getMemoryType());
+    } else {
+      // Setting the memory type.
+      // If either of the input and output needs to be a global memory
+      // tensor, both tensors should use global. Otherwise, use shared.
+      if (in_tv->isFusionInput() || in_tv->isFusionOutput() ||
+          out_tv->isFusionOutput()) {
+        in_tv->setMemoryType(MemoryType::Global);
+        out_tv->setMemoryType(MemoryType::Global);
+      } else {
+        in_tv->setMemoryType(MemoryType::Shared);
+        out_tv->setMemoryType(MemoryType::Shared);
+      }
+    }
+
+    scheduleScatterAllocationDomains(scatter);
+
+    // If there's a use of the scatter output, that must be the copy
+    // op inserted by insertCopyAfterScatter. It is not automatically
+    // scheduled as the propagation from the scatter output won't
+    // happen because the loop domain of the scatter output is not mapped
+    // with its logical domain.
+    if (!out_tv->uses().empty()) {
+      NVF_ERROR_EQ(out_tv->uses().size(), 1);
+      auto use_of_out = out_tv->uses().at(0);
+      NVF_ERROR(use_of_out->isA<LoadStoreOp>());
+      scheduleConstrainedTv(ir_utils::getTvOutput(use_of_out), {scatter_dim});
+    }
+  }
+
+  // Scatter-specific allocation domain scheduling
+  void scheduleScatterAllocationDomains(ScatterOp* scatter) {
+    auto in_tv = ir_utils::getTvInput(scatter);
+    auto out_tv = ir_utils::getTvOutput(scatter);
+
+    // For now, just use the logical domain.
+    if (!in_tv->hasAllocation()) {
+      in_tv->setAllocationDomain(in_tv->getLogicalDomain(), true);
+    }
+    if (!out_tv->hasAllocation()) {
+      out_tv->setAllocationDomain(out_tv->getLogicalDomain(), true);
+    }
```

Comment:
why do we need to explicitly set allocation domain at this point? This is after presegmentation passes, is there anything else that would alter allocation domain later?

Reply:
The scatter validation at the time of lowering asserts both input and output have an allocation domain. We could relax the validation by assuming that no allocation means that the lowering is free to choose whatever allocation domain. 

However, I think making it as explicit as possible, rather than having an implicit protocol between the scheduler and lowering, would help things more robust.